### ISSUE TITLE
[geometry] DOMMatrix constructor is a performance and code portability footgun

### ISSUE THREAD

**Issue by bzbarsky** [w3c/fxtf-drafts#346]
Consider what happens if one does:

  `new DOMMatrix(new DOMMatrix())`

(or in general passes a `DOMMatrix` instance to the `DOMMatrix` constructor).  The answer is that, since `DOMMatrix` is not iterable, it will get converted to a string of the form "matrix(stuff)", per https://drafts.fxtf.org/geometry/#dommatrixreadonly-stringification-behavior and then the constructor will either re-parse that string to create a `DOMMatrix` or throw, depending on whether it's running in a `Window` or a worker. This seems like a code portability (between windows and workers) footgun, as well as a performance footgun in the `Window` case.

Why, exactly, was the overload of the constructor that takes `DOMMatrixReadOnly` removed?  It seems to me that we should reinstate it and either make it consistently work (so people can use it) or consistently throw (so people know to not use it and use the `fromMatrix` thing instead)...  The current setup where it kinda works, but only in `Window` scopes doesn't seem great.

@zcorpan do you recall why this was set up like this?

@saschanaz

**Comment by zcorpan** [comment]
This was part of the change that introduced static methods to avoid overloading as much as possible. The string argument in the constructor was necessary for compat with web content using `WebKitCSSMatrix()`.

https://github.com/w3c/fxtf-drafts/commit/62b9cb9d5be4982d2a9cbf314e3a59efb8a68dd6

cc @domenic

**Comment by bzbarsky** [comment]
OK, but was the goal to remove "overloading" in the sense of using Web IDL overloads, or to reduce the set of things that could be passed in?  Because we can easily add `DOMMatrixReadOnly` to the union, right?

(Note that this situation is quite different from the one for `DOMPoint`, `DOMRect`, etc.)

**Comment by zcorpan** [comment]
The goal was to reduce WebIDL overloads.

That said, if people do `new DOMMatrix(new DOMMatrix())` (or variants), then I think it makes sense to add a `DOMMatrixReadOnly` overload. The only problem I think would be reduced consistency with other geometry APIs, but as you point out the current setup is a bit problematic.

**Comment by bzbarsky** [comment]
It's increased consistency on workers (where you can pass in a `DOMPoint` to the `DOMPoint` constructor and have it work, for example).

**Comment by zcorpan** [comment]
The new point would have x: NaN, y: 0, z: 0, w: 1, regardless of the passed point's properties, I believe. So it wouldn't throw, but also not really consistent?

**Comment by bzbarsky** [comment]
Hmm, so it would.  I thought for some reason that we still had a `DOMPointInit` constructor version for `DOMPoint`.  In that case, this is just an unfortunate interaction with the `toString` bit and the string constructor....

**Comment by saschanaz** [comment]
>That said, if people do new DOMMatrix(new DOMMatrix())

@bzbarsky should we add a use counter for this, or do you think we should add the overload regardless of use count?

If we are to add a counter, is it possible to add a counter for each overload on Firefox?

**Comment by bzbarsky** [comment]
You can add a counter to each overload in Firefox, yes.  You just have to do it manually, not via the IDL annotation.

I'm not sure a use counter is all that important here; the inconsistency between workers and mainthread bothers me more.

**Comment by tabatkins** [comment]
I'm curious - *why* did we want to reduce IDL overloads?

**Comment by zcorpan** [comment]
https://lists.w3.org/Archives/Public/public-fx/2015JanMar/0121.html
https://lists.w3.org/Archives/Public/public-fx/2015JanMar/0129.html

**Comment by tabatkins** [comment]
Ah, right! Thanks for digging up those references!

Okay, so it looks like we were just a bit overzealous here, then; we were already mentally committed to having the constructor take a DOMMatrixInit (which you can't distinguish from any other object), and so moved the entire chunk of argument-space over to a static method. Whereas we *could* have left DOMMatrixReadonly in the constructor's union, satisfying bz's concern in this thread and still fixing the original problem.

That would just mean that you could pass a literal DOMMatrix *either* in the constructor *or* the `.fromMatrix()` method and get the same result, but that seems fine.

**Comment by zcorpan** [comment]
Agree that that seems fine

**Comment by dirkschulze** [comment]
@bz @zcorpan @tabatkins So are you in agreement that we should re-add the overload for `DOMMatrixReadOnly` to the `DOMMatrix` constructor?

**Comment by tabatkins** [comment]
Yeah, I think it makes sense, and solves bz's very reasonable use-case.
